### ISSUE TITLE
MissingObjectException on BlobIterator

### ISSUE THREAD

**Issue by smola** [src-d/engine#357]
## Expected Behavior
This should work:

```sql
-- Top 100 tokens by repository count. Each token counted once per repository.
--
-- category: slow
SELECT
    token,
    COUNT(DISTINCT repository_id) AS repository_count
FROM (
    SELECT
        repository_id,
        EXPLODE(SPLIT(content, '[^_A-Za-z0-9]+')) AS token
    FROM blobs
    WHERE
        reference_name = 'refs/heads/HEAD' AND
        is_binary = false AND
        LENGTH(content) < 524288
) AS q
GROUP BY token
ORDER BY repository_count DESC
LIMIT 50
```

## Current Behavior
It crashes:

```
Job aborted due to stage failure: Task 2 in stage 12.0 failed 1 times, most recent failure: Lost task 2.0 in stage 12.0 (TID 843, localhost, executor driver): org.eclipse.jgit.errors.MissingObjectException: Missing unknown 871859241356f76917fdf7fcc3ba701b8c6bcbb5
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:163)
	at org.eclipse.jgit.lib.ObjectReader.open(ObjectReader.java:234)
	at tech.sourced.engine.iterator.BlobIterator$.readFile(BlobIterator.scala:90)
	at tech.sourced.engine.iterator.BlobIterator.mapColumns(BlobIterator.scala:56)
	at tech.sourced.engine.iterator.BlobIterator.mapColumns(BlobIterator.scala:16)
	at tech.sourced.engine.iterator.ChainableIterator.next(ChainableIterator.scala:111)
	at tech.sourced.engine.iterator.ChainableIterator.next(ChainableIterator.scala:16)
	at org.apache.spark.InterruptibleIterator.next(InterruptibleIterator.scala:40)
	at tech.sourced.engine.iterator.CleanupIterator.next(CleanupIterator.scala:36)
	at scala.collection.Iterator$$anon$12.next(Iterator.scala:445)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$JoinIterator.hasNext(Iterator.scala:212)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithKeys$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

## Possible Solution
Similar to the other `*Iterator` fixes for `MissingObjectException`

## Steps to Reproduce (for bugs)
1. Execute the query above on our staging cluster (`hdfs://hdfs-namenode/100-siva` ).
2. See the crash.

## Context
I was testing some SQL queries in our staging cluster.

## Your Environment (for bugs)
* Version used: engine-jupyter 0.5.5
* Operating System and version: k8s staging cluster
* Some needed resources to reproduce the problem: Did not identify offending siva file yet.

**Comment by bzz** [comment]
Hey, @carlosms after #336 would you mind also taking a look at this one - may be event possible to handle both with a same logic.

**Comment by smacker** [comment]
I also met this problem.

The interesting thing is:
1. I found it on ~1k dataset
2. Took first 500 files - can reproduce
3. Took first 250 files - can reproduce
4. Split into 2 datasets 125 files per each - both are processed successfully

**Comment by smola** [comment]
@smacker Try `getAllReferenceCommits` instead of `getCommits`. Maybe blobs are somewhere in the past history of HEAD and my query is somehow triggering that (although I think it shouldn't).

**Comment by smacker** [comment]
```
scala> engine.getRepositories.getReferences.filter('name === "refs/heads/HEAD").getAllReferenceCommits.getBlobs.filter('is_binary === false).select("content").map(_.size < 524288).count()
res8: Long = 40850
```

**Comment by carlosms** [comment]
I've tried to implement a fix in PR #358.
Although I haven't been able to reproduce it in spark shell:

```
# spark-shell --packages "tech.sourced:engine:0.5.5" --master="spark://p-spark-master:7077"

scala> import tech.sourced.engine._
import tech.sourced.engine._

scala> val engine = Engine(spark, "hdfs://hdfs-namenode/100-siva", "siva")
engine: tech.sourced.engine.Engine = tech.sourced.engine.Engine@d1b9dd7

scala> val sqlDF = spark.sql("""
     | SELECT
     |     token,
     |     COUNT(DISTINCT repository_id) AS repository_count
     | FROM (
     |     SELECT
     |         repository_id,
     |         EXPLODE(SPLIT(content, '[^_A-Za-z0-9]+')) AS token
     |     FROM blobs
     |     WHERE
     |         reference_name = 'refs/heads/HEAD' AND
     |         is_binary = false AND
     |         LENGTH(content) < 524288
     | ) AS q
     | GROUP BY token
     | ORDER BY repository_count DESC
     | LIMIT 50
     | """)
sqlDF: org.apache.spark.sql.DataFrame = [token: string, repository_count: bigint]

scala> sqlDF.show()
+-----+----------------+                                                        
|token|repository_count|
+-----+----------------+
|     |              85|
|    0|              78|
|    1|              77|
|   in|              75|
|    a|              75|
|   if|              75|
|  for|              75|
|   to|              74|
|   of|              74|
|    2|              74|
|  the|              73|
|  and|              73|
| this|              72|
|   be|              71|
| name|              71|
|  not|              70|
|   is|              70|
|    s|              70|
|   or|              69|
| from|              69|
+-----+----------------+
only showing top 20 rows

scala> sqlDF.count()
res0: Long = 50  
```
